wiki/python/ESTIMATOR_GUIDE.md

```diff
@@ -0,0 +1,866 @@
+
+# Access the CumlArrayDescriptorMeta value directly. No array conversion will occur
+print(my_est.__dict__["my_cuml_array_"])
+# Output: CumlArrayDescriptorMeta(input_type='cupy', values={'cuml': <cuml.common.array.CumlArray object at 0x7fd39174ae20>, 'numpy': array([ 0,  1,  1,  2,  2, -1, -1, ...
+
+# Values from CumlArrayDescriptorMeta can be specifically read
+print(my_est.__dict__["my_cuml_array_"].input_type)
+# Output: "cupy"
+
+# The input value can be accessed
+print(my_est.__dict__["my_cuml_array_"].get_input_value())
+# Output: CumlArray ...
+```
+
+### Estimator Methods
+
+#### Common Functionality
+
+All of these decorators perform the same basic steps with a few small differences. The common steps performed by each decorator is:
+
+1. Set `cuml.global_output_type = "mirror"`
+   1. When `"mirror"` is used as the global output type, that indicates we are in an internal cuML API call. The `CumlArrayDescriptor` keys off this value to change between internal and external functionality
+2. Set CuPy allocator to use RMM
+   1. This replaces the existing decorator `@with_cupy_rmm`
+   2. Unlike before, the CuPy allocator is only set once per API call
+3. Set the estimator input attributes. Can be broken down into 3 steps:
+   1. Set `_input_type` attribute
+   2. Set `target_dtype` attribute
+   3. Set `n_features` attribute
+4. Call the desired function
+5. Get the estimator output type. Can be broken down into 2 steps:
+   1. Get `output_type`
+   2. Get `output_dtype`
+6. Convert the return value
+   1. This will ultimately call `CumlArray.to_output(output_type=output_type, output_dtype=output_dtype)
+
+While the above list of steps may seem excessive for every call, most functions follow this general form, but may skip a few steps depending on a couple of factors. For example, Step #3 is necessary on functions that modify the estimator's estimated attributes, such as `fit()`, but is not necessary for functions like `predict()` or `transform()`. And Step #5/6 are only necessary when returning array-like objects and are ommitted when returning any other type.
+
+Functionally, you can think of these decorators equivalent to the following pseudocode:
+```python
+def my_func(self, X):
+   with cuml.using_ouput_type("mirror"):
+      with cupy.cuda.cupy_using_allocator(rmm.rmm_cupy_allocator):
+         # Set the input properties
+         self._set_base_attributes(output_type=X, n_features=X)
+
+         # Do actual calculation returning an array-like object
+         ret_val = self._my_func(X)
+
+         # Get the output type
+         output_type = self._get_output_type(X)
+
+         # Convert array-like to CumlArray
+         ret_val = input_to_cuml_array(ret_val, order="K").array
+
+         # Convert CumlArray to desired output_type
+         return ret_val.to_output(output_type)
+```
+
+Keep the above psuedocode in mind when working with these decorators since their goal is to replace many of these repetitive functions.
```

Comment:
```suggestion
Keep the above pseudocode in mind when working with these decorators since their goal is to replace many of these repetitive functions.
```